src/context-parser.js

```diff
@@ -287,8 +301,10 @@ FastParser.prototype.lookupChar = function(ch) {
 FastParser.prototype.contextualize = function(input, endsWithEOF) {
     var self = this, listeners = self.listeners, i = -1, lastState;
 
-    self.input = input;
-    self.inputLen = input.length;
+    // Perform input stream preprocessing
+    // Reference: https://html.spec.whatwg.org/multipage/syntax.html#preprocessing-the-input-stream
+    self.input = self.config.enableInputPreProcessing ? input.replace(reInputPreProcessing, function(m){return m[0] === '\r' ? '\n' : '\uFFFD';}) : input;
```

Comment:
does it mean that we are using batch replacement instead of char by char replacement?